torch/csrc/jit/frontend/source_range.h

```diff
@@ -4,37 +4,320 @@
+    if (tok.size() == 0) {
+      return 0;
+    }
+
+    if ((size() - start) < tok.size()) {
+      return std::string::npos;
+    }
+
+    Iterator begin = iter_for_pos(start);
+    Iterator end_iter = end();
+    size_t offset = start;
+    for (; begin != end_iter; ++begin, ++offset) {
+      if (*begin == tok[0]) {
+        auto mis = std::mismatch(begin, end_iter, tok.begin(), tok.end());
+        if (mis.second == tok.end()) {
+          // no mismatch, and second string (tok) is exhausted.
+          return offset;
+        }
+        if (mis.first == end_iter) {
+          // this str is exhausted but tok is not
+          return std::string::npos;
+        }
+      }
+    }
+    return std::string::npos;
+  }
+
+  StringCordView substr(size_t start, size_t size) const {
+    std::vector<c10::string_view> pieces;
+    std::vector<std::shared_ptr<std::string>> ownerships;
+    if (start >= this->size()) {
+      // out of bounds
+      return StringCordView();
+    }
+    if (start + size >= this->size()) {
+      size = this->size() - start;
+    }
+    Iterator begin = iter_for_pos(start);
+    Iterator end = iter_for_pos(start + size);
+
+    if (begin.line_ == end.line_) {
+      // same line
+      pieces.push_back(pieces_[begin.line_].substr(begin.pos_, size));
+    } else {
+      pieces.push_back(pieces_[begin.line_].substr(begin.pos_));
+
+      size_t last_line = pieces_.size();
+      if (end != this->end() && end.line_ < last_line) {
+        // end is within the string
+        last_line = end.line_;
+      }
+      for (size_t i = begin.line_ + 1; i < last_line; i++) {
+        pieces.push_back(pieces_[i]);
+      }
+      if (end != this->end()) {
+        pieces.push_back(pieces_[end.line_].substr(0, end.pos_));
+      }
+    }
+
+    // share ownership
+    std::copy(
+        owned_strings_.begin(),
+        owned_strings_.end(),
+        std::back_inserter(ownerships));
+
+    return StringCordView(std::move(pieces), std::move(ownerships));
+  }
+
+  char at(size_t index) const {
+    return *iter_for_pos(index);
+  }
+  char operator[](size_t index) const {
+    return at(index);
+  }
+
+  std::string str() const {
+    std::stringstream ss;
+    for (auto s : pieces_) {
+      ss << std::string(s);
+    }
+    return ss.str();
+  }
+
+  bool operator==(const std::string& rhs) {
+    if (size() != rhs.size()) {
+      return false;
+    }
+    auto res = std::mismatch(begin(), end(), rhs.begin(), rhs.end());
+    // both need to exhaust
+    return res.first == end() && res.second == rhs.end();
+  }
+
+  bool operator==(const StringCordView& rhs) {
+    if (size() != rhs.size()) {
+      return false;
+    }
+    auto res = std::mismatch(begin(), end(), rhs.begin(), rhs.end());
+    // both need to exhaust
+    return res.first == end() && res.second == rhs.end();
+  }
+
+  c10::string_view piece(size_t index) const {
+    return pieces_[index];
+  }
+
+  struct Iterator {
```

Comment:
Consider renaming the class, otherwise it would become `torch::jit::Iterator` which is too broad for what it covers.

Reply:
it's torch::jit::StringCordView::Iterator sicne it's nested inside the class.